server/src/main/scala/org/http4s/server/middleware/CSRF.scala

```diff
@@ -406,10 +407,26 @@ object CSRF {
       }
     }
 
+    def getMultipartToken: F[Option[String]] =
+      if (r.headers.get(headers.`Content-Type`).map(_.mediaType.isMultipart).getOrElse(false)) {
+        val token = multipart[G].decode(r, strict = true).value.flatMap {
```

Comment:
This seems unsafe to me: if we consume the body in a middleware, how can the application consume it a second time?  It will work for strict sources, but not for bytes streaming in off a socket.

But I see the same problem exists in `getFormToken`, and nobody has complained about that.  Is this something that definitely works in a live integration test, as opposed to just in unit tests?

Reply:
Regarding the `getFormToken`: We use CSRF and Forms (with a token field) in several apps and never had problems. However they only are regular forms.
On the other hand a client doing fancy javascript websocket streaming could simply send the appropriate header and circumvent the issue alltogether. The form parts are only called if no header is found.

Reply:
On a second thought... I think I'll try some integration test locally to avoid involving nasty suprises. After all the body parts are streams. :thinking: